### PR TITLE
Fix inconsistent number style + replaced `clientCulture` with currentCulture (code *was* inconsistent)

### LINKED ISSUE #2710: Inconsistent number style (group or thousand separators)
I'm not sure if this is dependent of someone's (browser) culture, but I'm seeing different number styles, which looks sloppy.

Current status
===
Home page has dots between thousands:
image

---

Stats page has dots between thousands:
image

---

Stats of a package has commas.
image

---

Packages overview has commas:
image

---

Version history of a package has **no** separator.. (difficult to read)

image

Resolution
----
Choose one format (preferred: browser or profile culture, please do use thousand separators), and use them everywhere! (maybe create a C# extension on int to prevent inconsistencies).

### PR DESCRIPTION
**Pull request by 304NotModified**
fixes #2710 

Created C# extensions for:

- `DetermineClientLocale` => `HttpRequestExtensions`
- Formatting number with client culture: `NumberExtensions.ToNuGetNumberString`

TODO:

- [x] unit test `UseClientCultureIfLanguageHeadersIsPresent` fails because I use `httpcontext.current`. Ways to fix this: 

  1. pass `Request` to the method everywhere instead of `httpcontext.current`, 
  2. mock `httpcontext.current`. 
  3. change unit test.
  4. remove the usage of `HttpRequest` in `ToString` (see 
 https://github.com/NuGet/NuGetGallery/pull/2716#issuecomment-145242154)   -  I think 1 isn't good, because we're changing code solely for an unit test. 
- [x] alignment of downloads at package versions (should by right)

### CODE REVIEW

**Comment by 304NotModified** [comment]
Todo

- [ ] alignment of downloads at package versions (should by right)

**Comment by 304NotModified** [comment]
I was thinking about this. I think the current solution isn't optimal. I should be:

request language => thread culture (UI + normal?) => toString methods with currentCulture.

This is more a .Net approach, and also, it fixes the tostrings of datetime etc. This also solves the unit test issue because we can set the current thread culture easily in an unit test. 

It's a more refactor step than this one. So I fix it now, or in a later PR. What do you think?

**Comment by 304NotModified** [comment]
Except the unit test,  this is done.

**Comment by 304NotModified** [comment]
Todo: 

- search :There are 43228 packages
Sorted by recent installs
Displaying results 1 - 20.

**Comment by 304NotModified** [comment]
Please give me feedback :)

**Comment by 304NotModified** [comment]
Note: I have created &  tested https://github.com/NuGet/NuGetGallery/pull/2716#issuecomment-145242154 in a separate branch. (not in this PR,  yet?) 
This will fix the unit test and also I have successfully removed clientCulture everywhere - which is an improvement IMO. Let me know if I should merge it into this PR.

File: src/NuGetGallery/Controllers/StatisticsController.cs [hunk 1]
```diff
@@ -41,62 +41,19 @@ public StatisticsController(IStatisticsService statisticsService, IAggregateStat
...
-            // if we fail to detect client locale from the Languages header, fall back to server locale
-            CultureInfo clientCulture = DetermineClientLocale() ?? CultureInfo.CurrentCulture;
+       
             return Json(
                 new
                 {
-                    Downloads = stats.Downloads.ToString("n0", clientCulture),
-                    UniquePackages = stats.UniquePackages.ToString("n0", clientCulture),
-                    TotalPackages = stats.TotalPackages.ToString("n0", clientCulture),
+                    Downloads = stats.Downloads.ToNuGetNumberString(),
```
Comment (maartenba) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: "ToLocalizedNumberString()" sounds more clear

**Comment by maartenba** [comment]
Added a minor comment to code (rename method so it's more clear what it does).

Regarding the unit test you can:
* safely mock the context
* I'd maybe pass the context into the extension methods at https://github.com/304NotModified/NuGetGallery/blob/304NotModified-2710-fix-inconsistent-number-style/src/NuGetGallery/Extensions/NumberExtensions.cs - it feels less clean when using the method but does make it perfectly clear that the context is needed. In fact I think passing in the Request object is enough - does not need the entire context

**Comment by 304NotModified** [comment]
> I'd maybe pass the context into the extension methods at https://github.com/304NotModified/NuGetGallery/blob/304NotModified-2710-fix-inconsistent-number-style/src/NuGetGallery/Extensions/NumberExtensions.cs - it feels less clean when using the method but does make it perfectly clear that the context is needed. In fact I think passing in the Request object is enough - does not need the entire context

Passing the Request wasn't always possible. I could write some more overloads, but everything was getting ugly. 

I only clean way was to implement my proposal of https://github.com/NuGet/NuGetGallery/pull/2716#issuecomment-145242154

Before executing an action (see `AppController.OnActionExecuting`), the `CurrentCulture` will be set by checking the Request.

I fully removed clientCulture. This is good because:

- `DateTimeExtensions` was using CurrentCulture already.
- The "official" way of .Net is using the CurrentCulture.
- Passing the ClientCulture was not nice. 
- This makes it unit testable. The unit test didn't even had to change :)
- Parsing the Userlanguage is now done once per request.

Status: all done

File: src/NuGetGallery/Controllers/StatisticsController.cs (hunk 1 above)
Comment (304NotModified) [inline src/NuGetGallery/Controllers/StatisticsController.cs]: I also don't like the name "ToNuGetNumberString". But it's consistent with `DateTimeExtensions`, like `ToNuGetShortDateTimeString`, `ToNuGetShortDateString` and `ToNuGetLongDateString`.

This is also mentioned in `NumberExtensions`:
image

**Comment by maartenba** [comment]
Thanks @304NotModified - will give this a thorough look tomorrow but looks sweet so far :)

**Comment by 304NotModified** [comment]
OK, I see the unit test still fails at appveyor, but works locally. Have to check that first. I will let you know when it's fixed.

- [ ] fix unit test

**Comment by 304NotModified** [comment]
the unit test has been fixed. See https://ci.appveyor.com/project/304NotModified/nugetgallery/build/1.0.20

**Comment by maartenba** [comment]
Merged into `dev` instead - https://github.com/NuGet/NuGetGallery/commit/440d22c411c526700f4cc06dcd18f0dbfbc36b87

Thanks @304NotModified for the contribution!

**Comment by 304NotModified** [comment]
No problem. But I preferred that I had created a new PR to "dev", because now the PR looks un-merged :(

**Comment by maartenba** [comment]
Ah no worries, that's GitHub I'm afraid.

**Pull request closed without merging**